src/Raven.Server/Documents/Queries/WhereFieldTokens.cs

```diff
@@ -60,30 +60,118 @@ private static string ExtractTokenValue(ValueToken valueToken, string queryText)
 
                 queryParameters.GetPropertyByIndex(index, ref propertyDetails);
 
-                switch (propertyDetails.Token)
+                return GetValueAndTypeFromJson(propertyDetails);
+            }
+
+            return (ExtractTokenValue(valueToken, queryText), Type);
+        }
+
+        private (List<string>, ValueTokenType) ExtractInParameters(ValueToken valueToken, string queryText, BlittableJsonReaderObject queryParameters)
+        {
+            Debug.Assert(Type == ValueTokenType.Parameter);
+
+            var property = new BlittableJsonReaderObject.PropertyDetails();
+
+            var parameterName = QueryExpression.Extract(queryText, valueToken);
+
+            var index = queryParameters.GetPropertyIndex(parameterName);
+
+            queryParameters.GetPropertyByIndex(index, ref property);
+
+            switch (property.Token)
+            {
+                case BlittableJsonToken.StartArray:
+                case BlittableJsonToken.StartArray | BlittableJsonToken.OffsetSizeByte:
+                case BlittableJsonToken.StartArray | BlittableJsonToken.OffsetSizeShort:
+                    if (Operator != OperatorType.In)
+                        throw new InvalidOperationException("Array parameter is supported only as a parameter of IN operator");
+
+                    return GetFlatValuesFromJsonArray((BlittableJsonReaderArray)property.Value);
+                default:
+                    var valueType = GetValueAndTypeFromJson(property);
+                    return (new List<string>{valueType.Value}, valueType.Type);
+            }
+        }
+
+        private static (string Value, ValueTokenType Type) GetValueAndTypeFromJson(BlittableJsonReaderObject.PropertyDetails property)
```

Comment:
We use strings to build Lucene query (terms). That code is being refactored and values will be retrieved when we really need them - once `QueryBuilder.BuildQuery` is called.